chain/client/src/view_client.rs

```diff
@@ -158,6 +157,28 @@ impl ViewClientActor {
         }
     }
 
+    fn get_block_hash_by_sync_checkpoint(
+        &mut self,
+        synchronization_checkpoint: &near_primitives::types::SyncCheckpoint,
+    ) -> Result<Option<CryptoHash>, Error> {
+        use near_primitives::types::SyncCheckpoint;
+
+        match synchronization_checkpoint {
+            SyncCheckpoint::Genesis => Ok(Some(self.chain.genesis().hash().clone())),
+            SyncCheckpoint::EarliestAvailable => {
+                let head_header_height = self.chain.head_header()?.height();
+                let mut chain_store_update = self.chain.mut_store().store_update();
+                let tail = chain_store_update.tail()?;
+                for height in tail..=head_header_height {
+                    if let Ok(block_hash) = chain_store_update.get_block_hash_by_height(height) {
+                        return Ok(Some(block_hash));
+                    }
+                }
+                Ok(None)
+            }
+        }
+    }
```

Comment:
This is the core of the PR (the rest is just renaming of `BlockIdOrFinality` to `BlockReference` as the original naming does not fit the extended semantic).

Reply:
I guess you want the following:
```
                for height in tail + 1..=head_header_height {
                    if let Ok(block_hash) = chain_store_update.get_block_hash_by_height(height) {
                        return Ok(Some(*chain_store_update.get_block_header(block_hash)?.prev_hash()));
                    }
                }
```
The reason is, the Tail is the position of where GC should continue working. The Tail is always equal of greater than the earliest Block in the Chain because in GC we're looking for the Block X (next one to the earliest one) and remove previous of X (the earliest one) - only this way Tries allow to delete data properly.